src/Lecture2.hs

```diff
@@ -211,7 +233,12 @@ The algorithm of merge sort is the following:
 [1,2,3]
 -}
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
+mergeSort []   = []
+mergeSort [x]  = [x]
+mergeSort list =
+  let middle = div (length list) 2
+      (left, right) = splitAt middle list
```

Comment:
You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.

Reply:
You are quite right. I thought about it and was implementing this sorting earlier in Erlang. I think there is not drawbacks  implement it this way 
```
mergeSort :: [Int] -> [Int]
mergeSort []   = []
mergeSort [x]  = [x]
mergeSort (x:xs) = merge [x] (mergeSort xs)
```

Reply:
That's a clever and correct solution 😅 
Unfortunately, it's not optimal asymptotically. Your solution has complexity `O(n ^ 2)` while optimal merge sort has `O(n log n)`. But the standard `sort` function in Haskell implements an idea similar to yours and improves it a bit to achieve even better performance 🐎